Main.hs

```diff
@@ -71,9 +72,7 @@ parseIf f =
       _          -> Nothing
 
 jsonNumber :: Parser JsonValue
-jsonNumber = f <$> spanP1 isDigit
-  where
-    f ds = JsonNumber $ read ds
+jsonNumber = JsonNumber <$> ((charP '-' $> (*(-1)) <|> pure id) <*> (read <$> spanP1 isDigit))
```

Comment:
I don't understand this piece of code. What is going on here?

Reply:
`((charP '-' $> (*(-1)) <|> pure id)` can be seen as
```Haskell
((charP '-' $> (*(-1)) -- if we manage to parse a '-', we discard it and send the "negate" function
  <|>
 (pure id)) -- if we don't see a '-' (minus) before the number, we send the id function
```

It then translates either of those two lines:
```Haskell
JsonNumber <$> (negate <$> (read <$> spanP1 isDigit))
-- or
JsonNumber <$> (id <$> (read <$> spanP1 isDigit))
```

But as `read :: String -> Double` handles negative numbers, we could just write a parser which reads a whole `Double` as text and read it.
```Haskell
jsonNumber = f <$> maybeSign <*> integer <*> maybeDecimal <*> maybeExponent
  where
    f sign i d expo = JsonNumber $ read (sign ++ i ++ "." ++ d ++ expo)
    maybeSign = stringP "-" <|> pure ""
    integer = spanP1 isDigit
    maybeDecimal = (charP '.' *> spanP1 isDigit) <|> pure "0"
    maybeESign = stringP "+" <|> maybeSign
    maybeExponent = ((++)
                    <$> ((:) <$> (charP 'e' <|> charP 'E') <*> maybeESign)
                    <*> integer) <|> pure ""
```

Not totaly sure about the implementation, but works on my tests ^^"

Reply:
Yes just as @ablanleuil explained if it reads a sign than negates or returns the same number otherwise if implementation move to Double than the above code should be added instead of this commit.